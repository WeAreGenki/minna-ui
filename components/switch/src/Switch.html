<!--
  INPUT SWITCH COMPONENT

  This component is intended to be an alternative to checkbox inputs in forms
  when there's a need to toggle something with an on and off state. From a UX
  perspective, it should be clearly understood that something is being turned on
  or off.

  An example of this is a switch to turn on or off sound in an interactive web
  app. On the other hand, you should use a checkbox when users are accustomed to
  seeing a checkbox such as an input to accept terms and conditions in a
  transaction flow. That said, most of the time this is a great option.

  USAGE:
    Simply import this component into your vue file and use it in your template
    as if it was a normal input; bind it to a vue model. Example:

      <Switch bind:value="switch"/>

    The default switch is the same size as buttons. If this is too big, you can
    add a "mini" property for a compact version. Example:

      <Switch bind:value="switch" mini/>

    It's possible to change the text from "ON"/"OFF" by passing the props textOn
    and textOff. If you change this you may need to modify the width too, which
    can be done via the --switch-width CSS variable. Example:

      <Switch bind:value="switch" textOn="YES" textOff="NO"/>

  TIPS:
    - It's a good idea to add a title attribute so desktop users can understand
      the elements purpose more clearly.
    - If following accessibility guidelines it's recommended to label the
      purpose of the switch by adding an `aria-label` or `aria-labelledby`
      attribute.
    - If used in a form this component won't send any data by default. If you
      wish to use it as a form input you'll need to bind the v-model to a
      hidden input.
    - The `disabled` attribute is not valid on <div> so we use a class.

  @format
-->

<!--
  TODO: Ability to drag the switch using mouse/touch in a high performance way.
-->

<div
  class="switch{mini ? ' switch-mini' : ''}{disabled !== undefined ? ' switch-disabled' : ''}{value ? ' switch-checked' : ''}"
  role="switch"
  tabindex="{disabled === undefined ? 0 : -1}"
  aria-checked="{value}"
  on:click="toggle()"
  on:keydown="handleKeyDown(event)"
>
  <div class="switch-slider"></div>
  <div class="switch-on">{textOn}</div>
  <div class="switch-off">{textOff}</div>
</div>

<script>
  export default {
    data: () => ({
      /* required bind prop */
      // value: true,

      /* optional props */
      textOn: 'ON',
      textOff: 'OFF',
      mini: false,

      /* optional native props */
      // disabled: false,
      // required: false, // FIXME: Add custom validation
    }),
    onstate({ changed, current, previous }) {
      if (!previous) return;

      if (changed.value) {
        this.fire('change', current.value);
      }
    },
    methods: {
      toggle() {
        if (this.get().disabled === undefined) {
          this.set({ value: !this.get().value });
        }
      },
      handleKeyDown(event) {
        // choose key with graceful fallback for old browsers
        switch (event.key || event.keyCode) {
          case 'Enter':
          case 13: // enter
          case ' ':
          case 'Spacebar':
          case 32:
            event.preventDefault(); // don't submit form (enter) or scroll page (space)
            this.toggle();
            break;
          default:
          // no matching key
        }
      },
    },
  };
</script>

<style type="text/postcss">
  @import './_switch.css';
  @import 'import.css';

  /* the switch purposely looks like .button and uses many of the button variables */
  .switch {
    position: relative;
    display: inline-flex;
    width: $switch-width;
    padding: $switch-padding;
    line-height: 1;
    color: $switch-off-text-color;
    text-align: center;
    cursor: pointer;
    background-color: $switch-off-background-color;
    border-radius: $button-radius;
    outline-offset: $outline-offset;
    box-shadow: $switch-inner-shadow, $switch-shadow;
    /* stylelint-disable-next-line plugin/no-low-performance-animation-properties */
    transition: background-color $switch-animate-out-speed ease-out;
    transition-delay: 0s;
    backface-visibility: hidden; /* promote into its own layer to avoid repaints */

    @if $switch-optimize {
      will-change: background-color;
    }

    &.switch-mini {
      width: $switch-mini-width;
      padding: $switch-mini-padding;
    }

    &:hover {
      outline: $input-hover-outline;
    }

    &:focus {
      outline: $input-focus-outline;
    }
  }

  .switch-checked {
    background-color: $switch-on-background-color;
    transition-delay: $switch-animate-in-speed;
    transition-duration: $switch-animate-in-speed;
  }

  .switch-on,
  .switch-off {
    flex-basis: calc($switch-width / 2);
  }

  .switch-on {
    color: $switch-on-text-color;

    .switch-disabled > & {
      color: $switch-disabled-text-color;
    }
  }

  .switch-slider {
    position: absolute;
    top: 0;
    bottom: 0;
    width: 50%;
    background: $button-background-color;
    border-radius: $button-radius;
    box-shadow: $switch-slider-inner-shadow;
    transition: transform $switch-animate-in-speed ease-out;

    @if $switch-optimize {
      will-change: transform;
    }

    .switch-checked > & {
      transform: translateX(100%);
    }

    .switch-mini > & {
      box-shadow: $switch-mini-slider-inner-shadow;
    }

    .switch:hover > &,
    .switch:focus > & {
      background: $switch-hover-slider-background-color;
    }

    /* stylelint-disable-next-line no-descending-specificity */
    .switch-disabled > & {
      box-shadow: none; /* inner shadow */
      transition: none;

      @if $switch-optimize {
        will-change: auto;
      }
    }

    /* stylelint-disable-next-line no-descending-specificity */
    .switch-disabled > &,
    .switch-disabled:hover > &,
    .switch-disabled:focus > & {
      background: $switch-disabled-slider-background-color;
    }
  }

  .switch-disabled {
    color: $button-disabled-text-color;
    cursor: not-allowed;
    background-color: $button-background-color;
    border: $switch-disabled-border;
    box-shadow: none; /* inner shadow */
    transition: none;

    @if $switch-optimize {
      will-change: auto;
    }

    /* stylelint-disable a11y/no-outline-none */
    &:hover,
    &:focus {
      outline: none;
    }
    /* stylelint-enable */
  }
</style>
