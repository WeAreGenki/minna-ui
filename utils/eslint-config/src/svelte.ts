/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

interface BlockAttributes {
  type: string;
}

export = {
  plugins: ['svelte3'],
  settings: {
    'svelte3/ignore-styles': (attr: BlockAttributes) =>
      attr.type === 'text/postcss',
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
