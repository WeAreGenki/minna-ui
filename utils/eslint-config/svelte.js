/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 *
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

'use strict';

/** @type {import('./types').ESLintConfig} */
module.exports = {
  plugins: ['svelte3'],
  settings: {
    /**
     * @param {import('./types').BlockAttributes} attr - Tag attributes.
     * @returns {boolean} If style block should be ignored.
     */
    'svelte3/ignore-styles': (attr) => attr.type === 'text/postcss',
    // TODO: Uncomment once false positives are handled better
    // 'svelte3/lint-template': true,
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
