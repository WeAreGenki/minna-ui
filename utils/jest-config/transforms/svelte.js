/**
 * Jest transform for compiling Svelte components into JavaScript.
 *
 * NOTE: `svelte` is not listed as a dependency so developers can use this
 * jest preset for non-svelte projects.
 */

'use strict';

const { compile } = require('svelte/compiler'); // eslint-disable-line import/no-extraneous-dependencies

/** @type {import('../types').JestTransformer} */
exports.process = (src, filename) => {
  // Strip out <style> tags to prevent errors when unable to parse PostCSS etc.
  const re = /<style[^>]*>[\S\s]*?<\/style>/g;
  const normalised = src.replace(re, '');

  const result = compile(normalised, {
    accessors: true, // Easy access to component props for tests
    css: false,
    dev: true,
    filename,
    format: 'cjs',
    preserveWhitespace: true,
  });

  const esInterop =
    'Object.defineProperty(exports, "__esModule", { value: true });';

  return {
    code: result.js.code + esInterop,
    map: result.js.map,
  };
};
