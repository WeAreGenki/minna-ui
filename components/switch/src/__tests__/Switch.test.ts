/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { tick } from 'svelte';
import Switch from '../Switch.svelte';

describe('Switch component', () => {
  it('renders correctly with no props', () => {
    expect.assertions(7);
    function wrapper(): void {
      const target = document.createElement('div');
      new Switch({ target });
      const el = target.querySelector('.switch')!;
      expect(el.getAttribute('tabindex')).toEqual('0');
      expect(el.getAttribute('disabled')).toBeNull();
      expect(el.getAttribute('required')).toBeNull();
      expect(document.querySelector('.switch-checked')).toBeNull();
      expect(document.querySelector('.switch-disabled')).toBeNull();
      expect(target.innerHTML).toMatchSnapshot();
    }
    expect(wrapper).not.toThrow();
  });

  it('renders with value true', () => {
    expect.assertions(2);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        value: true,
      },
      target,
    });
    expect(component.$$.ctx.value).toEqual(true);
    expect(target.querySelector('.switch-checked')).not.toBeNull();
  });

  it('renders with value false', () => {
    expect.assertions(2);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        value: false,
      },
      target,
    });
    expect(component.$$.ctx.value).toEqual(false);
    expect(target.querySelector('.switch-checked')).toBeNull();
  });

  it('renders with textOn/textOff props', () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        textOff: 'NO',
        textOn: 'YES',
      },
      target,
    });
    expect(component.$$.ctx.textOn).toEqual('YES');
    expect(component.$$.ctx.textOff).toEqual('NO');
    const switchOn = target.querySelector('.switch-on')!;
    const switchOff = target.querySelector('.switch-off')!;
    expect(switchOn.innerHTML).toEqual('YES');
    expect(switchOff.innerHTML).toEqual('NO');
  });

  it('renders with mini prop', () => {
    expect.assertions(2);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        mini: true,
      },
      target,
    });
    expect(component.$$.ctx.mini).toBeTruthy();
    expect(target.querySelector('.switch-mini')).not.toBeNull();
  });

  it('renders with disabled prop', () => {
    expect.assertions(3);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        disabled: true,
      },
      target,
    });
    expect(component.$$.ctx.disabled).toBeTruthy();
    expect(target.querySelector('.switch-disabled')).not.toBeNull();
    expect(target.querySelector('.switch')!.getAttribute('tabindex')).toEqual('-1');
  });

  it('toggles class when value changes', async () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        value: false,
      },
      target,
    });
    expect(component.$$.ctx.value).toEqual(false);
    expect(target.querySelector('.switch-checked')).toBeNull();
    component.$set({ value: true });
    await tick();
    expect(target.querySelector('.switch-checked')).not.toBeNull();
    component.$set({ value: false });
    await tick();
    expect(target.querySelector('.switch-checked')).toBeNull();
  });

  it('toggles value on click', async () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        value: true,
      },
      target,
    });
    expect(component.$$.ctx.value).toEqual(true);
    expect(target.querySelector('.switch-checked')).not.toBeNull();
    const el = target.querySelector<HTMLDivElement>('.switch')!;
    el.click();
    await tick();
    expect(component.$$.ctx.value).toEqual(false);
    expect(target.querySelector('.switch-checked')).toBeNull();
  });

  it('does not toggle value on click when disabled', async () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        disabled: true,
        value: true,
      },
      target,
    });
    expect(component.$$.ctx.value).toEqual(true);
    expect(target.querySelector('.switch-checked')).not.toBeNull();
    const el = target.querySelector<HTMLDivElement>('.switch')!;
    el.click();
    await tick();
    expect(component.$$.ctx.value).toEqual(true);
    expect(target.querySelector('.switch-checked')).not.toBeNull();
  });

  it('toggles on enter key press', async () => {
    expect.assertions(6);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        value: true,
      },
      target,
    });
    expect(component.$$.ctx.value).toEqual(true);
    expect(target.querySelector('.switch-checked')).not.toBeNull();
    const event1 = new KeyboardEvent('keydown', { key: 'Enter' });
    component.$$.ctx.handleKeyDown(event1);
    await tick();
    expect(component.$$.ctx.value).toEqual(false);
    expect(target.querySelector('.switch-checked')).toBeNull();
    // @ts-ignore - keyCode does actually exist!
    const event2 = new KeyboardEvent('keydown', { keyCode: 13 });
    component.$$.ctx.handleKeyDown(event2);
    await tick();
    expect(component.$$.ctx.value).toEqual(true);
    expect(target.querySelector('.switch-checked')).not.toBeNull();
  });

  it('toggles on spacebar key press', () => {
    expect.assertions(3);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        value: true,
      },
      target,
    });
    const event1 = new KeyboardEvent('keydown', { key: ' ' });
    component.$$.ctx.handleKeyDown(event1);
    expect(component.$$.ctx.value).toEqual(false);
    const event2 = new KeyboardEvent('keydown', { key: 'Spacebar' });
    component.$$.ctx.handleKeyDown(event2);
    expect(component.$$.ctx.value).toEqual(true);
    // @ts-ignore - keyCode does actually exist!
    const event3 = new KeyboardEvent('keydown', { keyCode: 32 });
    component.$$.ctx.handleKeyDown(event3);
    expect(component.$$.ctx.value).toEqual(false);
  });

  it('does not toggle on key press when disabled', () => {
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        disabled: true,
        value: true,
      },
      target,
    });
    const event = new KeyboardEvent('keydown', { key: 'Enter' });
    component.$$.ctx.handleKeyDown(event);
    expect(component.$$.ctx.value).toEqual(true);
  });

  it('does not toggle on invalid key press', () => {
    expect.assertions(5);
    const target = document.createElement('div');
    const component = new Switch({
      props: {
        value: true,
      },
      target,
    });
    const spy = jest.spyOn(component.$$.ctx, 'toggle');
    const event1 = new KeyboardEvent('keydown', { key: 'Escape' });
    component.$$.ctx.handleKeyDown(event1);
    expect(component.$$.ctx.value).toEqual(true);
    // @ts-ignore - keyCode does actually exist!
    const event2 = new KeyboardEvent('keydown', { keyCode: 0 });
    component.$$.ctx.handleKeyDown(event2);
    expect(component.$$.ctx.value).toEqual(true);
    const event3 = new KeyboardEvent('keydown', { key: '' });
    component.$$.ctx.handleKeyDown(event3);
    expect(component.$$.ctx.value).toEqual(true);
    const event4 = new KeyboardEvent('keydown');
    component.$$.ctx.handleKeyDown(event4);
    expect(component.$$.ctx.value).toEqual(true);
    expect(spy).not.toHaveBeenCalled();
    spy.mockRestore();
  });
});
