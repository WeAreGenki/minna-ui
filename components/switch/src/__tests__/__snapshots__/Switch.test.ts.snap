// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component renders correctly with no props 1`] = `"<div class=\\"switch   \\" role=\\"switch\\" tabindex=\\"0\\" aria-checked=\\"false\\"><div class=\\"switch-slider\\"></div> <div class=\\"switch-on\\">ON</div> <div class=\\"switch-off\\">OFF</div></div>"`;
